[MRG] Predict probabilities in forest using sample weight

qcw171717 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
Forest ensembles take the mean of class weights from each tree, which may sometimes be problematic. For example, in a trained forest at prediction time, the input sample falls into a leaf with 5 samples from the training set in the first tree while it falls into a leaf with 5000 samples from the training set in the second tree. With the current implementations, this difference in sample size from the training examples are not taken into consideration when making predictions. The method **predict_proba_with_sample_weight** will allow all forest ensembles to predict class probabilities with each tree's vote weighted by how many times it's seen similar data points in the training set (number of samples in the leaf the input sample falls into). Users can also call the predict function in classification problems with sample_weight set to True, this will call the function **predict_proba_with_sample_weight** in place of **predict_proba**.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
test

Commit:
study proba

Commit:
study proba

Commit:
testing sample weighted proba

Commit:
check normalizer

Commit:
fixed return normalizer in decision trees

Commit:
fixed return normalizer in decision trees

Commit:
fixed sample weight calc

Commit:
fixed sample weight calc

Commit:
fixed sample weight calc

Commit:
sample weight proba and predict version 0

Commit:
sample weight proba and predict version 0 test

Commit:
sample weight proba and predict version 0 test

Commit:
sample weight proba and predict version 0 test

Commit:
sample weight proba and predict version 0 test

Commit:
sample weight proba and predict version 0 test

Commit:
added function descriptions

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into sample_weight_proba

jnothman commented:
Not able to review your pr yet, but: we will not add a new method. We can
either add a way to allow the user to configure this change in
predict_proba behaviour, or consider it a bug fix and just change it
outright.

Commit:
deleted extra methods and added parameter options

Commit:
cleaned up remaining predict_with_proba method

Commit:
fixed bug in not using sample weight

Commit:
changed parameter name in method descr and methods

Commit:
fixed bugs due to remaining misnomers

Commit:
fixed linting issues

Commit:
fixed linting issues test2

qcw171717 commented:
> Not able to review your pr yet, but: we will not add a new method. We can either add a way to allow the user to configure this change in predict_proba behaviour, or consider it a bug fix and just change it outright.

Thank you for the comment and suggestion, I deleted any additional methods I added in the PR, and added a use_sample_proba parameter in the predict_proba method, defaulted to False to used the current behaviour, and when configured to True, uses the added behaviour described in this PR. Please let me know if this is acceptable.

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into sample_weight_proba

Commit:
fixed unknown target warning

Commit:
merged master

Commit:
added test

Commit:
fixing lint problem

Commit:
added parallel test

thomasjpfan commented:
I am closing this PR because of https://github.com/scikit-learn/scikit-learn/pull/17160#issuecomment-1196022107
